python/tests/test_ownership_variable.py

```diff
@@ -0,0 +1,543 @@
+    inner['x', 2] = -3
+    inner.unit = 's'
+    a[3] = -4
+    assert sc.identical(
+        outer, make_variable(make_variable([-1, -2, -3, 3, 4], unit='s')))
+    assert sc.identical(inner, make_variable([-1, -2, -3, 3, 4], unit='s'))
+    np.testing.assert_array_equal(a, [0, 1, 2, -4, 4])
+
+    # Assignment creates a new inner Variable.
+    outer.value = make_variable(np.arange(5, 8), unit='kg')
+    outer.value['x', 0] = -5
+    assert sc.identical(outer,
+                        make_variable(make_variable([-5, 6, 7], unit='kg')))
+    assert sc.identical(inner, make_variable([-1, -2, -3, 3, 4], unit='s'))
+
+    # TODO this causes and infinite loop and segfaults at the moment
+    # outer.value = outer
+    # assert sc.identical(outer, outer.value)
+
+
+def test_own_var_scalar_nested_get():
+    # Variables are shared when nested.
+    outer = make_variable(make_variable(np.arange(5)))
+    inner = outer.value
+    a = inner.values
+    outer.value['x', 0] = -1
+    outer.value.values[1] = -2
+    inner['x', 2] = -3
+    a[3] = -4
+    expected = [-1, -2, -3, -4, 4]
+    assert sc.identical(outer, make_variable(make_variable(expected)))
+    assert sc.identical(inner, make_variable(expected))
+    np.testing.assert_array_equal(a, expected)
+
+    # Assignment writes the values into the inner Variable.
+    outer.value = make_variable(np.arange(5, 8))
+    outer.value['x', 0] = -5
+    inner['x', 1] = -10
+    assert sc.identical(outer, make_variable(make_variable([-5, -10, 7])))
+    assert sc.identical(inner, make_variable([-5, -10, 7]))
+    np.testing.assert_array_equal(a, [-1, -2, -3, -4, 4])
+
+
+def test_own_var_scalar_nested_copy():
+    # Depth of copies of variables can be controlled.
+    outer = make_variable(make_variable(np.arange(5)))
+    inner = outer.value
+    a = inner.values
+    outer_copy = copy(outer)
+    outer_deepcopy = deepcopy(outer)
+    outer_methcopy = outer.copy(deep=False)
+    outer_methdeepcopy = outer.copy(deep=True)
+
+    outer.value['x', 0] = -1
+    outer.value.values[1] = -2
+    inner['x', 2] = -3
+    a[3] = -4
+    modified = make_variable(make_variable([-1, -2, -3, -4, 4]))
+    original = make_variable(make_variable([0, 1, 2, 3, 4]))
+    assert sc.identical(outer, modified)
+    assert sc.identical(outer_copy, modified)
+    assert sc.identical(outer_deepcopy, original)
+    assert sc.identical(outer_methcopy, modified)
+    assert sc.identical(outer_methdeepcopy, original)
+
+
+def test_own_var_scalar_nested_str_get():
+    # Variables are shared when nested.
+    outer = make_variable(make_variable(np.array(['abc', 'def'])))
+    inner = outer.value
+    a = inner.values
+    # Assignment writes the values into the inner Variable.
+    outer.value = make_variable(np.array(['ghi', 'jkl', 'mno']))
+    outer.value['x', 0] = sc.scalar('asd')
+    inner['x', 1] = sc.scalar('qwe')
+    assert sc.identical(
+        outer, make_variable(make_variable(np.array(['asd', 'qwe', 'mno']))))
```

Comment:
Is this doing what it should? I would have expected that after `out.value = other_var`, `inner` would still hold a reference to the original, and not affect `outer.value` and more?

Reply:
This is consistent with what happens in `test_own_var_scalar_nested_get`, see our comments above.

I am not sure whether that is good or bad.

Reply:
Hmm, it is a bit confusing. On the one hand I would expect `obj.value = other` to just point the `value` property to `other`. On the other hand, we cannot in general, since values are part of `Variable` and not available as a PYthon object. So probably consistent and ok?